Add message if JavaScript is disabled

mkaz opened the pull request:
## Description

Adds a noscript and message if Javascript is disabled.

This also required adding a :not(.no-js) to the fullscreen body class
which hides/shows the rest of the wp-admin chrome.



## How has this been tested?

* Disable Javascript using console settings 
1. Prior to change confirm the page is blank.
2. After change you should see the message

<img width="711" alt="screen shot 2018-11-08 at 11 23 34 am" src="https://user-images.githubusercontent.com/45363/48222398-60be4480-e349-11e8-9c7c-ac342315b53d.png">


## Types of changes

See #2719 

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [X] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [X] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

mkaz committed:
Add message if Javascript is disabled

Adds a noscript and message if Javascript is disabled.

This also required adding a :not(.no-js) to the fullscreen body class
which hides/shows the rest of the wp-admin chrome.

mkaz committed:
PHP linting

ocean90 commented on gutenberg.php:
```diff
@@ -29,6 +29,13 @@ function the_gutenberg_project() {
+		<noscript>
+			<div class="error" style='margin-top:32px'><p>
+			<?php
+			echo __( 'The block editor requires Javascript to be enabled.', 'gutenberg' );
```
`echo __()` => `_e()`

mkaz replied:
The `echo __()` function is used in numerous spots in gutenberg.php, I just mirrored it to stay consistent within the same file.

mkaz replied:
Thumbs down is not really a productive response.
I've switched each of the occurrences to the shorter function.

ocean90 commented on gutenberg.php:
```diff
@@ -29,6 +29,13 @@ function the_gutenberg_project() {
 	?>
 	<div class="block-editor gutenberg">
+		<noscript>
+			<div class="error" style='margin-top:32px'><p>
```
The admin bar has different heights depending on the viewport.

mkaz replied:
The margin was simply to give some spacing from the bottom of the admin bar, so the error message isn't squished up against the top.

mkaz committed:
12 bytes saved

swissspidy commented on packages/edit-post/src/components/fullscreen-mode/style.scss:
```diff
@@ -1,4 +1,4 @@
-body.is-fullscreen-mode {
```
See also #11677

youknowriad reviewed (approved):
LGTM 👍

michelleweber commented:
I always like to give a specific next step in an error message when possible. In this case, that would be something like "The Block Editor requires Javascript. Go to your browser's settings and enable Javascript before using it. Learn more." 

Ideally the "learn more" is a link to instructions re: how to enable JS, but if we don't have that, just delete it.

chrisvanpatten commented on gutenberg.php:
```diff
@@ -29,6 +29,13 @@ function the_gutenberg_project() {
+		<noscript>
+			<div class="error" style='margin-top:32px'><p>
+			<?php
+			_e( 'The block editor requires Javascript to be enabled.', 'gutenberg' );
```
JavaScript should be CamelCased!

```suggestion
			_e( 'The block editor requires JavaScript to be enabled.', 'gutenberg' );
```

chrisvanpatten reviewed (changes requested):
Small typo!

mkaz committed:
Fix camel-case, update with Copy Review

mkaz commented:
@michelleweber I've updated with your change, but after revisiting the bug, I think [the suggestion here](https://github.com/WordPress/gutenberg/issues/2719#issuecomment-437435084) by @afercia might be better, adding a to link to the Classic Editor plugin 

The user probably either has JavaScript disabled on purpose, or it not supported by their device, so the plugin instructions might be more helpful.

michelleweber commented:
Works for me. As long as we give the user some kind of next step, we're good -- I leave it to you to decide what the actual best next step is!

mkaz committed:
Switch message to include link to Classic Editor plugin

swissspidy commented:
I know this is now getting a bit more complex, but can we add some more conditions whether the Classic Editor plugin is already installed and whether it's active? Perhaps also whether the user has the capability to install it.

At least a direct link to the classic editor if the plugin is already active would be nice.

mkaz committed:
Fix link not be clickable

Add position absolute and higher z-index to make link clickable.
The block-editor pane was positioned above the inserted link.

h/t @georgehotelling

afercia commented on packages/edit-post/src/components/fullscreen-mode/style.scss:
```diff
@@ -1,4 +1,4 @@
-body.is-fullscreen-mode {
+body.is-fullscreen-mode:not(.no-js) {
```
rather than `:not(.no-js)` I'd simply use `body.js.is-fullscreen-mode`

mkaz replied:
Thank you @afercia !

mkaz commented:
@swissspidy I know, so much effort on a rare edge case, but might as well do it properly.

I did consider that about if the plugin is installed, however the user should not get to this page since it is a Gutenberg page and if the plugin is installed they will go straight to the Classic Editor.

I also just had to push up some CSS z-index positioning stuff to make it clickable, no such thing as an quick fix.

afercia commented:
@mkaz thanks for working on this!

Seems it's not possible to actually click on the link because the `#editor` container overlays the notice 🙂 

@jasmussen is printing out by default `is-fullscreen-mode` via PHP really necessary? I guess there are JS ways to set it before everything is rendered.

mkaz committed:
Use existence of .js class instead of :not(.no-js)

h/t to @afercia

mkaz commented:
@afercia can you make sure you are updated to the latest PR, I just pushed up a fix to address the issue of it being clickable. Also, just updated with your CSS suggestion.

swissspidy commented:
What if the plugin is installed but not active?

afercia commented:
The link is clickable now, thanks!

GaryJones commented on gutenberg.php:
```diff
@@ -28,6 +28,16 @@
 	global $post_type_object;
 	?>
+	<noscript>
+		<div class="error" style='position:absolute;top:32px;z-index:40'><p>
```
Why use single quotes for the style attribute?

GaryJones commented on gutenberg.php:
```diff
@@ -28,6 +28,16 @@
+		<div class="error" style='position:absolute;top:32px;z-index:40'><p>
+		<?php
+			printf(
+				__( 'The Block Editor requires JavaScript, please try the <a href="%s">Classic Editor plugin</a>.', 'gutenberg' ),
```
This will read better as two sentences.
"The Block Editor requires JavaScript. Please try the Classic Editor plugin."

gziolo commented:
> The Block Editor requires JavaScript. Please try the Classic Editor plugin.

I think that this message is valid in the context of WordPress core. I'm not convinced that it is the way to go when Gutenberg plugin is activated. @mcsf @youknowriad @mtias do you agree?

mkaz committed:
Add conditional for plugin vs core

Two different messages if trying to access block editor via the
plugin or via core.

mkaz commented:
@mcsf @gziolo - I've updated based on your suggestions.

gziolo commented:
Thanks @mkaz for updating 👍

mcsf commented on gutenberg.php:
```diff
@@ -28,6 +28,25 @@
+		<?php
+			// Using Gutenberg as Plugin
+			if ( is_plugin_active( 'gutenberg/gutenberg.php' ) ) {
+				$current_url = esc_url( $_SERVER['REQUEST_URI'] . "&classic-editor" );
```
This (concatenating `"&classic-editor"`) will break if `REQUEST_URI` doesn't carry prior query arguments. Should use `add_query_arg` or equivalent.

mkaz replied:
I've updated, though I'm not sure if that can happen on the editor page.

mcsf replied:
> I've updated, though I'm not sure if that can happen on the editor page.

By default, Add New Post leads to `/wp-admin/post-new.php` and Add New Page leads to `/wp-admin/post-new.php?post_type=page`. Both situations are to be expected, so `add_query_arg` is definitely necessary.

mkaz committed:
Use add_query_arg()

chrisvanpatten commented:
@mkaz Oh sorry, I thought the review would have been auto-dismissed after you incorporated my suggestion. Get that fixed up, GitHub!

Approving in a sec :)

mkaz commented:
Thanks @chrisvanpatten I marked as resolved too, I do think it's a Github bug.
